### ISSUE TITLE
printf output truncated

### ISSUE THREAD

**Issue by agrover** [iovisor/bpftrace#2129]
Printing a network packet only includes the first eight bytes. Adding a `printf("")` before the real printf seems to help.

### What reproduces the bug?

```
#include <linux/skbuff.h>

kprobe:__netif_receive_skb
{
  $skb = (struct sk_buff*)arg0;
  $buf = buf($skb->data, (int64)($skb->len));

  if ($skb->len == 1288) {
    printf("ip SKB len %d +%r+\n", $skb->len, $buf);
  }
}
```

### bpftrace --info
System
  OS: Linux 5.15.15-cloudflare-kasan-2022.1.7 #1 SMP Mon Sep 27 00:00:00 UTC 2010
  Arch: x86_64

Build
  version: v0.14.0
  LLVM: 11.0.1
  ORC: v2
  foreach_sym: yes
  unsafe uprobe: no
  bfd: yes
  bpf_attach_kfunc: yes
  bcc_usdt_addsem: yes
  bcc bpf_attach_uprobe refcount: yes
  bcc library path resolution: yes
  libbpf: yes
  libbpf btf dump: yes
  libbpf btf dump type decl: yes
  libdw (DWARF support): no

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: yes
  get_boot_ns: yes
  dpath: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf (depends on Build:libbpf): yes
  map batch (depends on Build:libbpf): yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  iter:task: yes
  iter:task_file: yes

**Comment by jsitnicki** [comment]
Steps to reproduce:
```
$ cat t1.bt
kprobe:__netif_receive_skb
{
  $skb = (struct sk_buff*)arg0;
  $buf = buf($skb->data, $skb->len);

  if ($skb->len == 1288) {
    printf("ip SKB len %d +%r+\n", $skb->len, $buf);
  }
}
$ sudo bpftrace t1.bt -c '/usr/bin/ping -q -c1 -s 1260 -p cf 127.0.0.1'
Attaching 1 probe...
PING 127.0.0.1 (127.0.0.1) 1260(1288) bytes of data.

--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.053/0.053/0.053/0.000 ms
ip SKB len 1288 +E\x00\x05\x08\*@\x00+
ip SKB len 1288 +E\x00\x05\x08\+\x00\x00+


$ cat t2.bt
kprobe:__netif_receive_skb
{
  $skb = (struct sk_buff*)arg0;
  $buf = buf($skb->data, $skb->len);

  printf("");

  if ($skb->len == 1288) {
    printf("ip SKB len %d +%r+\n", $skb->len, $buf);
  }
}
$ sudo bpftrace t2.bt -c '/usr/bin/ping -q -c1 -s 1260 -p cf 127.0.0.1'
Attaching 1 probe...
PING 127.0.0.1 (127.0.0.1) 1260(1288) bytes of data.

--- 127.0.0.1 ping statistics ---
ip SKB len 1288 +E\x00\x05\x08\xa5\x81@\x00@\x01\x92q\x7f\x00\x00\x01\x7f\x00\x00\x01\x08\x00\xd7\x1c,M\x00\x01*\xb2\xdda\x00\x00\x00\x00\xfd\x95\x04\x00\x00\x00\x00\x00\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf+
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.098/0.098/0.098/0.000 ms
ip SKB len 1288 +E\x00\x05\x08\xa5\x82\x00\x00@\x01\xd2p\x7f\x00\x00\x01\x7f\x00\x00\x01\x00\x00\xdf\x1c,M\x00\x01*\xb2\xdda\x00\x00\x00\x00\xfd\x95\x04\x00\x00\x00\x00\x00\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf+


$
```

**Comment by fbs** [comment]
Sadly cannot reproduce

```
$ sudo ./v0.14/src/bpftrace 1.bt -c '/usr/bin/ping -q -c1 -s 1260 -p cf 127.0.0.1'
Attaching 1 probe...
PING 127.0.0.1 (127.0.0.1) 1260(1288) bytes of data.

ip SKB len 1288 +E\x00\x05\x08Di@\x00@\x01\xf3\x89\x7f\x00\x00\x01\x7f\x00\x00\x01\x08\x00d\xc7\x00\x0d\x00\x01s0\xe7a\x00\x00\x00\x00J\xad\x03\x00\x00\x00\x00\x00\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf+
ip SKB len 1288 +E\x00\x05\x08Dj\x00\x00@\x013\x89\x7f\x00\x00\x01\x7f\x00\x00\x01\x00\x00l\xc7\x00\x0d\x00\x01s0\xe7a\x00\x00\x00\x00J\xad\x03\x00\x00\x00\x00\x00\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf+
```

which distro is this on?

**Comment by jsitnicki** [comment]
I ran the repro on a Debian Bullseye VM running an 5.15 LTS based kernel.
Perhaps the kernel version makes a difference? Here's the translated BPF prog that I'm seeing loaded:
```
vagrant@bullseye:~$ sudo bpftool prog dump xlated name __netif_receive
   0: (bf) r6 = r1
   1: (b7) r9 = 0
   2: (73) *(u8 *)(r10 -96) = r9
   3: (7b) *(u64 *)(r10 -104) = r9
   4: (7b) *(u64 *)(r10 -112) = r9
   5: (7b) *(u64 *)(r10 -120) = r9
   6: (7b) *(u64 *)(r10 -128) = r9
   7: (7b) *(u64 *)(r10 -136) = r9
   8: (7b) *(u64 *)(r10 -144) = r9
   9: (7b) *(u64 *)(r10 -152) = r9
  10: (7b) *(u64 *)(r10 -160) = r9
  11: (79) r8 = *(u64 *)(r6 +112)
  12: (bf) r7 = r8
  13: (07) r7 += 112
  14: (bf) r1 = r10
  15: (07) r1 += -88
  16: (b7) r2 = 4
  17: (bf) r3 = r7
  18: (85) call bpf_probe_read_kernel#-62128
  19: (73) *(u8 *)(r10 -80) = r9
  20: (73) *(u8 *)(r10 -81) = r9
  21: (73) *(u8 *)(r10 -82) = r9
  22: (73) *(u8 *)(r10 -83) = r9
  23: (73) *(u8 *)(r10 -84) = r9
  24: (73) *(u8 *)(r10 -85) = r9
  25: (73) *(u8 *)(r10 -86) = r9
  26: (73) *(u8 *)(r10 -87) = r9
  27: (73) *(u8 *)(r10 -72) = r9
  28: (73) *(u8 *)(r10 -73) = r9
  29: (73) *(u8 *)(r10 -74) = r9
  30: (73) *(u8 *)(r10 -75) = r9
  31: (73) *(u8 *)(r10 -76) = r9
  32: (73) *(u8 *)(r10 -77) = r9
  33: (73) *(u8 *)(r10 -78) = r9
  34: (73) *(u8 *)(r10 -79) = r9
  35: (73) *(u8 *)(r10 -64) = r9
  36: (73) *(u8 *)(r10 -65) = r9
  37: (73) *(u8 *)(r10 -66) = r9
  38: (73) *(u8 *)(r10 -67) = r9
  39: (73) *(u8 *)(r10 -68) = r9
  40: (73) *(u8 *)(r10 -69) = r9
  41: (73) *(u8 *)(r10 -70) = r9
  42: (73) *(u8 *)(r10 -71) = r9
  43: (73) *(u8 *)(r10 -56) = r9
  44: (73) *(u8 *)(r10 -57) = r9
  45: (73) *(u8 *)(r10 -58) = r9
  46: (73) *(u8 *)(r10 -59) = r9
  47: (73) *(u8 *)(r10 -60) = r9
  48: (73) *(u8 *)(r10 -61) = r9
  49: (73) *(u8 *)(r10 -62) = r9
  50: (73) *(u8 *)(r10 -63) = r9
  51: (73) *(u8 *)(r10 -48) = r9
  52: (73) *(u8 *)(r10 -49) = r9
  53: (73) *(u8 *)(r10 -50) = r9
  54: (73) *(u8 *)(r10 -51) = r9
  55: (73) *(u8 *)(r10 -52) = r9
  56: (73) *(u8 *)(r10 -53) = r9
  57: (73) *(u8 *)(r10 -54) = r9
  58: (73) *(u8 *)(r10 -55) = r9
  59: (73) *(u8 *)(r10 -40) = r9
  60: (73) *(u8 *)(r10 -41) = r9
  61: (73) *(u8 *)(r10 -42) = r9
  62: (73) *(u8 *)(r10 -43) = r9
  63: (73) *(u8 *)(r10 -44) = r9
  64: (73) *(u8 *)(r10 -45) = r9
  65: (73) *(u8 *)(r10 -46) = r9
  66: (73) *(u8 *)(r10 -47) = r9
  67: (73) *(u8 *)(r10 -32) = r9
  68: (73) *(u8 *)(r10 -33) = r9
  69: (73) *(u8 *)(r10 -34) = r9
  70: (73) *(u8 *)(r10 -35) = r9
  71: (73) *(u8 *)(r10 -36) = r9
  72: (73) *(u8 *)(r10 -37) = r9
  73: (73) *(u8 *)(r10 -38) = r9
  74: (73) *(u8 *)(r10 -39) = r9
  75: (73) *(u8 *)(r10 -24) = r9
  76: (73) *(u8 *)(r10 -25) = r9
  77: (73) *(u8 *)(r10 -26) = r9
  78: (73) *(u8 *)(r10 -27) = r9
  79: (73) *(u8 *)(r10 -28) = r9
  80: (73) *(u8 *)(r10 -29) = r9
  81: (73) *(u8 *)(r10 -30) = r9
  82: (73) *(u8 *)(r10 -31) = r9
  83: (b7) r1 = 64
  84: (61) r9 = *(u32 *)(r10 -88)
  85: (2d) if r1 > r9 goto pc+1
  86: (b7) r9 = 64
  87: (73) *(u8 *)(r10 -88) = r9
  88: (07) r8 += 200
  89: (bf) r1 = r10
  90: (07) r1 += -168
  91: (b7) r2 = 8
  92: (bf) r3 = r8
  93: (85) call bpf_probe_read_kernel#-62128
  94: (79) r3 = *(u64 *)(r10 -168)
  95: (bf) r1 = r10
  96: (07) r1 += -87
  97: (bf) r2 = r9
  98: (85) call bpf_probe_read_kernel#-62128
  99: (79) r1 = *(u64 *)(r10 -88)
 100: (7b) *(u64 *)(r10 -160) = r1
 101: (79) r1 = *(u64 *)(r10 -80)
 102: (7b) *(u64 *)(r10 -152) = r1
 103: (79) r1 = *(u64 *)(r10 -72)
 104: (7b) *(u64 *)(r10 -144) = r1
 105: (79) r1 = *(u64 *)(r10 -64)
 106: (7b) *(u64 *)(r10 -136) = r1
 107: (79) r1 = *(u64 *)(r10 -56)
 108: (7b) *(u64 *)(r10 -128) = r1
 109: (79) r1 = *(u64 *)(r10 -48)
 110: (7b) *(u64 *)(r10 -120) = r1
 111: (79) r1 = *(u64 *)(r10 -40)
 112: (7b) *(u64 *)(r10 -112) = r1
 113: (79) r1 = *(u64 *)(r10 -32)
 114: (7b) *(u64 *)(r10 -104) = r1
 115: (71) r1 = *(u8 *)(r10 -24)
 116: (73) *(u8 *)(r10 -96) = r1
 117: (bf) r1 = r10
 118: (07) r1 += -88
 119: (b7) r2 = 4
 120: (bf) r3 = r7
 121: (85) call bpf_probe_read_kernel#-62128
 122: (61) r1 = *(u32 *)(r10 -88)
 123: (55) if r1 != 0x508 goto pc+46
 124: (b7) r1 = 0
 125: (7b) *(u64 *)(r10 -88) = r1
 126: (7b) *(u64 *)(r10 -80) = r1
 127: (7b) *(u64 *)(r10 -72) = r1
 128: (7b) *(u64 *)(r10 -64) = r1
 129: (7b) *(u64 *)(r10 -56) = r1
 130: (7b) *(u64 *)(r10 -48) = r1
 131: (7b) *(u64 *)(r10 -40) = r1
 132: (7b) *(u64 *)(r10 -32) = r1
 133: (7b) *(u64 *)(r10 -24) = r1
 134: (7b) *(u64 *)(r10 -16) = r1
 135: (7b) *(u64 *)(r10 -8) = r1
 136: (bf) r1 = r10
 137: (07) r1 += -168
 138: (b7) r2 = 4
 139: (bf) r3 = r7
 140: (85) call bpf_probe_read_kernel#-62128
 141: (61) r1 = *(u32 *)(r10 -168)
 142: (7b) *(u64 *)(r10 -80) = r1
 143: (79) r1 = *(u64 *)(r10 -160)
 144: (7b) *(u64 *)(r10 -72) = r1
 145: (79) r1 = *(u64 *)(r10 -152)
 146: (7b) *(u64 *)(r10 -64) = r1
 147: (79) r1 = *(u64 *)(r10 -144)
 148: (7b) *(u64 *)(r10 -56) = r1
 149: (79) r1 = *(u64 *)(r10 -136)
 150: (7b) *(u64 *)(r10 -48) = r1
 151: (79) r1 = *(u64 *)(r10 -128)
 152: (7b) *(u64 *)(r10 -40) = r1
 153: (79) r1 = *(u64 *)(r10 -120)
 154: (7b) *(u64 *)(r10 -32) = r1
 155: (79) r1 = *(u64 *)(r10 -112)
 156: (7b) *(u64 *)(r10 -24) = r1
 157: (79) r1 = *(u64 *)(r10 -104)
 158: (7b) *(u64 *)(r10 -16) = r1
 159: (71) r1 = *(u8 *)(r10 -96)
 160: (73) *(u8 *)(r10 -8) = r1
 161: (18) r2 = map[id:8]
 163: (bf) r4 = r10
 164: (07) r4 += -88
 165: (bf) r1 = r6
 166: (18) r3 = 0xffffffff
 168: (b7) r5 = 88
 169: (85) call bpf_perf_event_output#-61280
 170: (b7) r0 = 0
 171: (95) exit
vagrant@bullseye:~$
```

**Comment by viktormalik** [comment]
Running the debug build fails on an assertion:
```
bpftrace: /usr/include/llvm/IR/Instructions.h:1207: void llvm::ICmpInst::AssertOK(): Assertion `getOperand(0)->getType() == getOperand(1)->getType() && "Both operands to ICmp instruction are not of the same type!"' failed.
```

That could to be the issue here - a problem occurs when the second param of `buf` is not a 64-bit int.

I have a fix here:
https://github.com/viktormalik/bpftrace/tree/buf-codegen-fix

Could you try if it helps in your case?

**Comment by viktormalik** [comment]
This is fixed on my system by #2217. Could you verify that this is the case for you, too? Thanks!

**Comment by jsitnicki** [comment]
Confirmed. Thank you!

```
vagrant@bullseye:~$ sudo bpftrace t1.bt -c '/usr/bin/ping -q -c1 -s 1260 -p cf 127.0.0.1'
Attaching 1 probe...
PING 127.0.0.1 (127.0.0.1) 1260(1288) bytes of data.

--- 127.0.0.1 ping statistics ---
ip SKB len 1288 +E\x00\x05\x08d\xe1@\x00+
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.035/0.035/0.035/0.000 ms
ip SKB len 1288 +E\x00\x05\x08d\xe2\x00\x00+


vagrant@bullseye:~$ sudo ./bpftrace/build/src/bpftrace t1.bt -c '/usr/bin/ping -q -c1 -s 1260 -p cf 127.0.0.1'
Attaching 1 probe...
PING 127.0.0.1 (127.0.0.1) 1260(1288) bytes of data.

--- 127.0.0.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.034/0.034/0.034/0.000 ms
ip SKB len 1288 +E\x00\x05\x08h\x0c@\x00@\x01\xcf\xe6\x7f\x00\x00\x01\x7f\x00\x00\x01\x08\x00\x05\xad\x10\x1c\x00\x01\x9f\xa4\x8bb\x00\x00\x00\x00\xc2C\x0a\x00\x00\x00\x00\x00\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf+
ip SKB len 1288 +E\x00\x05\x08h\x0d\x00\x00@\x01\x0f\xe6\x7f\x00\x00\x01\x7f\x00\x00\x01\x00\x00\x0d\xad\x10\x1c\x00\x01\x9f\xa4\x8bb\x00\x00\x00\x00\xc2C\x0a\x00\x00\x00\x00\x00\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf\xcf+


vagrant@bullseye:~$
```
